lib/spack/spack/package_base.py

```diff
@@ -2555,7 +2555,7 @@ class PackageStillNeededError(InstallError):
     """Raised when package is still needed by another on uninstall."""
 
     def __init__(self, spec, dependents):
-        super().__init__("Cannot uninstall %s" % spec)
+        super().__init__("Cannot uninstall %s, needed by %s" % (spec, dependents))
```

Comment:
Please use an f-string (https://docs.python.org/3/tutorial/inputoutput.html) since we are transitioning the core to using them when we make changes.

Reply:
(Optional) IMO I would prefer to see a more concise representation of the dependents.  For example, instead of:

```
spack.package_base.PackageStillNeededError: Cannot uninstall mpich@=3.0.4%gcc@=10.2.1~debug build_system=generic arch=test-debian6-core2, needed by mpileaks@=2.3%gcc@=10.2.1~debug~opt+shared+static build_system=generic arch=test-debian6-core2 ^[deptypes=build,link] callpath@=1.0%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link] dyninst@=8.2%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link] libdwarf@=20130729%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link] libelf@=0.8.13%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link virtuals=mpi] mpich@=3.0.4%gcc@=10.2.1~debug build_system=generic arch=test-debian6-core2, callpath@=1.0%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link] dyninst@=8.2%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link] libdwarf@=20130729%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link] libelf@=0.8.13%gcc@=10.2.1 build_system=generic arch=test-debian6-core2 ^[deptypes=build,link virtuals=mpi] mpich@=3.0.4%gcc@=10.2.1~debug build_system=generic arch=test-debian6-core2
```

I would rather see something along the lines:

```
spack.package_base.PackageStillNeededError: Cannot uninstall mpich@=3.0.4%gcc@=10.2.1~debug build_system=generic arch=test-debian6-core2, needed by mpileaks@2.3, callpath@1.0
```

What do you think?

Reply:
I agree, though in this case that would have hidden the issue that I was trying to debug since it was about the dependencies and multiple build dependencies with same version getting installed. I think that, since these are installed packages, your approach would work if we can also attach the hash, since then it is unambiguous which package is referred to.

Reply:
I don't have a good way to trigger this exception, but I now get something like this:
```
>>> specs = spack.store.STORE.db.query("zlib")
>>> from spack.package_base import PackageStillNeededError
>>> PackageStillNeededError(specs[0], specs)
spack.package_base.PackageStillNeededError("Cannot uninstall zlib@=1.3.1%gcc@=11.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64_v3 /azgz7kt, needed by ['zlib@=1.3.1 /azgz7kt']",None)
```
so a string that looks like:
```
Cannot uninstall zlib@=1.3.1%gcc@=11.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64_v3 /azgz7kt, needed by ['zlib@=1.3.1 /azgz7kt']
```
What do you think of this?

Reply:
I agree that having the hash is very useful.  And it is definitely tighter than the original.  Tweaking a unit test I get:

```
spack.package_base.PackageStillNeededError: Cannot uninstall mpich@=3.0.4%gcc@=10.2.1~debug build_system=generic arch=test-debian6-core2 /ddxyfch, needed by ['mpileaks@=2.3 /qadhdym', 'callpath@=1.0 /7kvnioh']
```

Works for me if it works for you.